File: advanced-programming/higher-order-functions/memoize/memoize.rb
require 'net/http'
require 'uri'

def memoize(meth)
  original = method(meth)
  cache = {}

  define_method(meth) do |*args, **kwargs, &block|
    key = args
    return cache[key] if cache.key?(key)

    cache[key] = original.call(*args, **kwargs, &block)
  end
end

memoize def fetch(url)
  uri = URI(url)
  Net::HTTP.get(uri)
end

memoize def fib(n)
  return n if n <= 1
  fib(n - 1) + fib(n - 2)
end

puts fib(35)
puts fetch("https://google.com")

File: advanced-programming/higher-order-functions/memoize/memoize.py
import urllib.request

def memoize(fn):
    def inner(arg):
        if arg in cache:
            return cache[arg]
        res = fn(arg)
        cache[arg] = res
        return res
    return inner

cache = {}

def fetch(url):
    with urllib.request.urlopen(url) as response:
        content = response.read().decode('utf-8')
        return content

@memoize
def fib(n):
    if n <= 1:
        return n
    return fib(n - 1) + fib(n - 2)


if __name__ == '__main__':
    # fetch = memoize(fetch)
    # print(fetch('http://google.com')[:80])
    # print(fetch('http://google.com')[80:160])
    # print(fetch('http://google.com')[160:240])
    # print(fetch('http://google.com')[240:320])

    print(fib(35))
